Update repository-management.md with proposed guidelines

annezazu opened the pull request:
In light of today's core editor chat, this change aims to incorporate proposed guidelines from @youknowriad to make creating the changelog easier: 

https://make.wordpress.org/core/2020/05/27/editor-chat-summary-27th-may-2020/

The wording is slightly refined and made more explicit. I also included additional updates wanted from design.

annezazu committed:
Update repository-management.md

In light of today's core editor chat, this change aims to incorporate proposed guidelines from @youknowriad to make creating the changelog easier: https://make.wordpress.org/core/2020/05/27/editor-chat-summary-27th-may-2020/

aduth commented on docs/contributors/repository-management.md:
```diff
@@ -70,6 +70,14 @@ Gutenberg follows a feature branch pull request workflow for all code and docume
+
+- When working on experimental screens and features, apply the `[Type] Experimental` label instead of `Feature`, `Enhancement`, etc.
+- When working on new features to technical packages (scripts, create-block, adding  react hooks, etc), apply the `[Type] New API`  label instead of `Feature`, `Enhancement`, etc.
+- When fixing a bug or making an enhancement to an internal tool used in the project, apply the `[Type] Build Tooling` instead of `Bugs`, `Enhancement`, etc
```
For some of these, I wonder if there should be better disambiguation between _the type of change_ and _the area of focus of the change_. There are already "[Package]"-prefixed labels, and it makes sense to consider "a bug fix within the DOM package". Couldn't it also make sense to think of "an enhancement to build tooling", etc.

I guess I think "Build Tooling" is maybe not a "type" in the same way that "Bug Fix" or "Enhancement" could be.

At the same time though:

- It's not a package
- For the changelog tool, it should still be captured under the grouping "Build Tooling"

Maybe one of the following:

- Remove "[Type]" prefix and have some hard-coded awareness of this label in the changelog tool
- Use "[Feature]" prefix instead
- Try to push more of these "Build Tooling" into packages, then have mappings of packages which have their own groupings (or subgroupings) in release notes
   - To this point, if you look at the [8.2](https://make.wordpress.org/core/2020/05/27/whats-new-in-gutenberg-27-may/) or [8.0](https://make.wordpress.org/core/2020/04/29/whats-new-in-gutenberg-29-april/) release notes, you can see lists for specific packages like "wordpress/env", "wordpress/scripts". So there is some precedent.

Note: I think the same applies to the "[Type] Experimental" guidance above. Experimental is also interesting because you can see clearly in the [8.2 release notes](https://make.wordpress.org/core/2020/05/27/whats-new-in-gutenberg-27-may/) that there's a desire to group these by the specific experiment. Maybe there needs to be an additional level of categorization (labeling) here.

Finally: #22601 does add some prioritization to how these type labels are assigned, but because it uses the same mechanism for ordering groups, it's maybe not always what we'd expect ("Bug Fixes" would be prioritized over "Experiments" or "Build Tooling").

cc @youknowriad

youknowriad replied:
I  think "Bug Fixes", "Enhancement", "Feature"  should always have less priority than the remaining types. (We might want to clarify that on the tool).

II think it's fine to have  two types in that case: "type bug" and "type build tooling". "type experimental" and "type enhancement"...

Whether we should name these differently than "types", I don't know, maybe but I'd like to avoid complexifying the system and the tool. I'd be ok with leaving types even if not 100% accurate.

annezazu replied:
Reading this through, I'm struggling to know what to update here (if anything) :D This is likely my own ignorance. It seems like there might be split thoughts around [type] vs [feature] and some of this feedback lands more about the changelog tool itself rather than this update.  What am I missing? What can I need to do to keep this moving forward?

aduth replied:
I don't think it warrants any immediate change to the document, at least given how things stand today.

As you note, it's more a separate discussion if and how we could better disambiguate "types of changes" from "focus of changes". I still think there's some inconsistencies (e.g. categorizing "focus" by "[Package]", "[Focus]", or non-prefixed labels), but not something that needs to be settled here.

aduth commented on docs/contributors/repository-management.md:
```diff
@@ -70,6 +70,14 @@ Gutenberg follows a feature branch pull request workflow for all code and docume
+- When working on experimental screens and features, apply the `[Type] Experimental` label instead of `Feature`, `Enhancement`, etc.
+- When working on new features to technical packages (scripts, create-block, adding  react hooks, etc), apply the `[Type] New API`  label instead of `Feature`, `Enhancement`, etc.
+- When fixing a bug or making an enhancement to an internal tool used in the project, apply the `[Type] Build Tooling` instead of `Bugs`, `Enhancement`, etc
+- In PR titles, instead of describing the code change done to fix an issue, consider referring to the actual bug being fixed instead. For example: instead of saying "Check for nullable object in component", it would be preferable to say "Fix editor breakage when clicking the copy block button". 
```
"PR" is a fairly common abbreviation, but it might not be universally understood, so it wouldn't hurt to spell it out in its complete form:

```suggestion
- In pull request titles, instead of describing the code change done to fix an issue, consider referring to the actual bug being fixed instead. For example: instead of saying "Check for nullable object in component", it would be preferable to say "Fix editor breakage when clicking the copy block button". 
```

aduth reviewed (approved):
Did you consider if it might be valuable to also include a link to this guidance in the pull request template?

https://github.com/WordPress/gutenberg/blob/master/.github/PULL_REQUEST_TEMPLATE.md

annezazu committed:
Update docs/contributors/repository-management.md

Co-authored-by: Andrew Duthie <andrew@andrewduthie.com>

annezazu commented:
@aduth Yes! That's next on my list to get to once this was approved (working in chunks to make sure there's consensus).

annezazu commented on docs/contributors/repository-management.md:
```diff
@@ -70,6 +70,14 @@ Gutenberg follows a feature branch pull request workflow for all code and docume
 2. Make your changes, testing thoroughly.
 3. Commit your changes when you’re happy with them, and push the branch.
 4. Open your pull request.
+5. Label and name your pull request appropriately (see below).
```
This likely needs to be adjusted to the following since non contributors won't have the ability to self label their own PRs: 

5. If you have the ability to do so, label and name your pull request appropriately (see below).

aduth commented:
> @aduth Yes! That's next on my list to get to once this was approved (working in chunks to make sure there's consensus).

Ah, fair enough 😄  I see #22681 now.

youknowriad commented on docs/contributors/repository-management.md:
```diff
@@ -70,6 +70,14 @@ Gutenberg follows a feature branch pull request workflow for all code and docume
 4. Open your pull request.
+5. Label and name your pull request appropriately (see below).
+
+For labeling and naming pull requests, here are guidelines to follow that make compiling the changelog more efficient and organized:
```
how can we both have these guidelines or maybe "recommendation" and also avoid having to force contributors to learn dozens of things in order to try opening a PR.

Should we clarify that it applies essentially for regular contributors?
Should we make it clear that it's totally fine to make mistakes and learn as you do?

annezazu replied:
Just updated this section to make the language more of a recommendation specifically for regular contributors and to be more "mistake friendly" :) Resolving this but let me know if the new language still doesn't sit right.

annezazu committed:
Update repository-management.md

Adjustments to language in light of feedback from @youknowriad to make it clear that it's okay to make mistakes and that these guidelines are more to consider rather than get perfectly.
